src/device/r4300/fpu.h

```diff
@@ -74,33 +85,188 @@ M64P_FPU_INLINE void set_rounding(uint32_t fcr31)
     }
 }
 
-M64P_FPU_INLINE void cvt_s_w(uint32_t fcr31, const int32_t* source, float* dest)
+M64P_FPU_INLINE void fpu_reset_cause(uint32_t* fcr31)
+{
+    (*fcr31) &= ~FCR31_CAUSE_BITS;
+}
+
+M64P_FPU_INLINE void fpu_reset_exceptions()
+{
+    feclearexcept(FE_ALL_EXCEPT);
+}
+
+M64P_FPU_INLINE int fpu_check_exceptions(uint32_t* fcr31)
+{
+    int fexceptions;
+
+    fexceptions = fetestexcept(FE_ALL_EXCEPT) & FE_ALL_EXCEPT;
+
+    if (fexceptions & FE_DIVBYZERO)
+    {
+        (*fcr31) |= FCR31_CAUSE_DIVBYZERO_BIT;
+        (*fcr31) |= FCR31_FLAG_DIVBYZERO_BIT;
+    }
+    if (fexceptions & FE_INEXACT)
+    {
+        (*fcr31) |= FCR31_CAUSE_INEXACT_BIT;
+        (*fcr31) |= FCR31_FLAG_INEXACT_BIT;
+    }
+    if (fexceptions & FE_UNDERFLOW)
+    {
+        (*fcr31) |= FCR31_CAUSE_UNDERFLOW_BIT;
+        (*fcr31) |= FCR31_FLAG_UNDERFLOW_BIT;
+    }
+    if (fexceptions & FE_OVERFLOW)
+    {
+        (*fcr31) |= FCR31_CAUSE_OVERFLOW_BIT;
+        (*fcr31) |= FCR31_FLAG_OVERFLOW_BIT;
+    }
+    if (fexceptions & FE_INVALID)
+    {
+        (*fcr31) |= FCR31_CAUSE_INVALIDOP_BIT;
+        (*fcr31) |= FCR31_FLAG_INVALIDOP_BIT;
+    }
+
+    return 0; // TODO: exceptions
+}
+
+M64P_FPU_INLINE void fpu_check_input_float(uint32_t* fcr31, const float* value)
 {
-    set_rounding(fcr31);
+    switch (fpclassify(*value))
+    {
+    default:
+    case FP_SUBNORMAL: // TODO
+        return;
+    case FP_NAN:
+        (*fcr31) |= FCR31_CAUSE_INVALIDOP_BIT;
+        (*fcr31) |= FCR31_FLAG_INVALIDOP_BIT;
+        break;
+    }
+}
+
+M64P_FPU_INLINE void fpu_check_input_double(uint32_t* fcr31, const double* value)
+{
+    switch (fpclassify(*value))
+    {
+    default:
+    case FP_SUBNORMAL: // TODO
+        return;
+    case FP_NAN:
+        (*fcr31) |= FCR31_CAUSE_INVALIDOP_BIT;
+        (*fcr31) |= FCR31_FLAG_INVALIDOP_BIT;
+        break;
+    }
+}
+
+M64P_FPU_INLINE void fpu_check_output_float(uint32_t* fcr31, const float* value)
+{
+    switch (fpclassify(*value))
+    {
+    case FP_SUBNORMAL:
+        (*fcr31) |= FCR31_CAUSE_UNDERFLOW_BIT;
+        (*fcr31) |= FCR31_FLAG_UNDERFLOW_BIT;
+        (*fcr31) |= FCR31_CAUSE_INEXACT_BIT;
+        (*fcr31) |= FCR31_FLAG_INEXACT_BIT;
+        break;
+
+    case FP_NAN:
+        break;
+
+    default:
+        break;
+    }
+}
+
+M64P_FPU_INLINE void fpu_check_output_double(uint32_t* fcr31, const double* value)
+{
+    switch (fpclassify(*value))
+    {
+    case FP_SUBNORMAL:
+        (*fcr31) |= FCR31_CAUSE_UNDERFLOW_BIT;
+        (*fcr31) |= FCR31_FLAG_UNDERFLOW_BIT;
+        (*fcr31) |= FCR31_CAUSE_INEXACT_BIT;
+        (*fcr31) |= FCR31_FLAG_INEXACT_BIT;
+        break;
+
+    case FP_NAN:
+        break;
+
+    default:
+        break;
+    }
+}
+
+M64P_FPU_INLINE void cvt_s_w(uint32_t* fcr31, const int32_t* source, float* dest)
+{
+    set_rounding(*fcr31);
+
+    fpu_reset_cause(fcr31);
+    fpu_reset_exceptions();
+
     *dest = (float)*source;
+
+    fpu_check_exceptions(fcr31);
```

Comment:
I don't think this should be included. I haven't tested, but checking for exceptions like this must create a fairly severe performance penalty.

I think this kind of thing should be wrapped in compile time defines like:
```
#if CHECK_EXCEPTIONS
...
#endif
```

This kind of thing is really only useful for homebrew development. There is no need to turn mupen64plu into Ares. Ares has its strengths, and mupen64plus has its strengths, but worsening the performance of mupen64plus, without providing any benefit to commercial games, is not playing to mupen64plus's strengths.

Reply:
> I don't think this should be included. I haven't tested, but checking for exceptions like this must create a fairly severe performance penalty.

I disagree, the performance seems largely fine in games with my ryzen 3600, yes this will have a slight performance impact but I feel that the added accuracy (at the moment setting the correct flag/cause bits) is worth it overall.

> This kind of thing is really only useful for homebrew development. There is no need to turn mupen64plu into Ares. Ares has its strengths, and mupen64plus has its strengths, but worsening the performance of mupen64plus, without providing any benefit to commercial games, is not playing to mupen64plus's strengths.

I disagree, mupen64plus is a nintendo 64 emulator and should mimick the real hardware behavior as much as possible, and i.e other emulators (like Project64) are working towards getting n64-systemtest to pass most tests aswell, so if we don't improve our accuracy, mupen64plus will lack behind the other emulators, which I feel is a bad thing.

Reply:
Yes, but many mupen64plus users aren't using a ryzen 3600, they are using raspberry pi's. I'm just saying that this additional level of accuracy should be made optional, so as to not affect performance, since it doesn't affect the emulation of commercial games.

Reply:
Personally, I prefer that any resource intensive settings can be switched at will. I like the precise emulation, but also the flexibility of being able to use it on any PC designed for an office (low performance PC).
If that's not possible in the immediate future, I'll have to agree with Logan.

Reply:
I agree - changes which have significant performance regressions should be removable with compile-time options. Lots of people run M64+ on low-spec hardware.

Reply:
Fixed with https://github.com/mupen64plus/mupen64plus-core/pull/1001/commits/e1f425156921321c164e24cda7ba83b16f736938, now there's a `ACCURATE_FPU=1` option in the Makefile, which is disabled by default.